---
title: "Day 2 - Module 8a: Introduction to DESeq2"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            theme: readable
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: true
            fig_caption: true
            keep_md: false
---

<!--- Allow the page to be wider --->
<style>
    body .main-container {
        max-width: 1200px;
    }
</style>
> # Objectives
> * Broad introduction to DESeq2 & why it is widely used for differential expression comparisons

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE,warning=FALSE}
library(DESeq2)
load("RunningData.RData")
```

# Tools for Differential Gene Expression analysis

As discussed during the seminar, a common application for RNA-seq is to testing for differential expression between conditions, using statistical approaches that are appropriate for the type of data we expect from biological data.

While there are several tools that can be used for differential expression comparisons, we use [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) in our analysis today. DESeq2 is one of two tools, along with [EdgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html), considered ['best practice'](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-91) for differential expression, as both tools apply similar methods that account for the distributions we expect to see for RNA-seq and are fairly stringent in calling differentially expressed genes.

Additionally, `DESeq2` also has an 
[excellent vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 
from Love, Anders, and Huber from which our workflow is partially adapted and is a good resource when analyzing your own data
(see also: [Love, Anders, and Huber. _Genome Biology_. 2014.](https://doi.org/10.1186/s13059-014-0550-8)).


<details>
  <summary> Additional resources regarding statistical testing and tool comparison for RNA-seq data
To learn more about statistical testing and what distributions best model the behavior of RNA-seq data, a good resource is this [EdX lecture by Rafael Irizarry](https://www.youtube.com/watch?v=HK7WKsL3c2w&feature=youtu.be) or this [lecture by Kasper Hansen](https://www.youtube.com/watch?v=C8RNvWu7pAw). Another helpful guide is this [Comparative Study for Differential Expression Analysis by Zhang et al.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0103207) from 2014.

</details>

We previously loaded several libraries into our R session, we can check the tools documentation this out using the `?` operator.
```{r CheckDocumentaion}
?`DESeq2-package`
```

## DESeq2 assumptions and requirements

A key assumption is that since a vast number of RNAs are represented in each sample, the probability of counting a particular transcript is small. In addition, we expect that the mean expression of any given gene is less than the variance (or spread) of expression across different samples. 

<details>
  <summary> Biological versus technical variance & statistical thinking around RNA-seq data
For most experiments biological variance is much greater than technical variance, especially if [best practices](https://www.txgen.tamu.edu/faq/rna-isolation-best-practices/) for [quality RNA isolation](https://www.biocompare.com/Bench-Tips/128790-Four-Tips-for-Perfecting-RNA-Isolation/) are followed (including DNase treatment!).

> **Note**: If you are working with cell lines and would like more detail about replicates for *in vitro* studies, this [archived page](https://web-archive-org.proxy.lib.umich.edu/web/20170807192514/http://www.labstats.net:80/articles/cell_culture_n.html) could be helpful.

For a walk through of mean versus variance in RNA-seq data, this [blog post by David Tang](https://davetang.org/muse/2012/04/14/variance-in-rna-seq-data/) has a nice practical walk through. A more technical talk by Gordon Smyth is summarized in this [blog post by Peter Hicky](https://www.peterhickey.org/2011/11/23/bioinf-seminar-gordon-smyth/) and described in detail in [Symth's voom paper](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29)
</details>

Since variance is a key to both the approach used in DESeq2, if you try to compare treamtment groups with less than **two** replicates, DESeq2 will give you an error.

### Replicates in RNA-seq experiments

A question we are frequently asked is "How many replicates do I need?"

As the below image (from the [Harvard Chan Bioinformatics Core](https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html) training materials) summarizes, there is often more contributing to observed gene expression than we expect. 


![](./images/de_variation.png)

The general goal of differential expression analysis to attempt to separate the “interesting” biological contributions from the “uninteresting” technical or extraneous contributions that either cannot be or were not controlled inthe experimental design.

A related aspect is how much to consider sequencing depth versus additional replicates. This figure shared by Illumina in their technical talks is helpful to understand the relative importance of sequencing depth versus number of replicates.


![](./images/de_replicates_img.png)


Generally, sequencing depth is important for accurately measuring genes that are expressed at different abundances. For the human and mouse genomes, the general recommendation is 30-40 million reads per sample if measuring the ~20,000  protein-coding genes (i.e.: polyA library prep) to measure both highly expressed (common) and more lowly expresssed (rarer) transcripts. However, as the image above shows, replicates make a bigger impact on detecting differentially expressed genes (DEGs)

DESeq2 has some additional requirments, including the type of data expected as an input, which we will discuss later.

---

# Sources Used
* HBC DGE setup: https://hbctraining.github.io/DGE_workshop/lessons/01_DGE_setup_and_overview.html
* HBC Count Normalization: https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html
* DESeq2 standard vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
* DESeq2 beginners vignette: https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf
* Bioconductor RNA-seq Workflows: https://www.bioconductor.org/help/course-materials/2015/LearnBioconductorFeb2015/B02.1_RNASeq.html

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
#Hidden code block to write out data for knitting
save.image(file = "RunningData.RData")
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
